plugins/woocommerce/src/Internal/ProductQueryFilters/FilterDataProvider.php

```diff
@@ -0,0 +1,364 @@
+		$product_query_sql = $query->request;
+
+		remove_filter( 'posts_clauses', array( $this->filter_clauses_generator, 'add_query_clauses' ), 10 );
+		remove_filter( 'posts_pre_query', '__return_empty_array' );
+
+		$attributes_to_count_sql = 'AND term_taxonomy.taxonomy IN ("' . esc_sql( wc_sanitize_taxonomy_name( $attribute_to_count ) ) . '")';
+		$attribute_count_sql     = "
+			SELECT COUNT( DISTINCT posts.ID ) as term_count, terms.term_id as term_count_id
+			FROM {$wpdb->posts} AS posts
+			INNER JOIN {$wpdb->term_relationships} AS term_relationships ON posts.ID = term_relationships.object_id
+			INNER JOIN {$wpdb->term_taxonomy} AS term_taxonomy USING( term_taxonomy_id )
+			INNER JOIN {$wpdb->terms} AS terms USING( term_id )
+			WHERE posts.ID IN ( {$product_query_sql} )
+			{$attributes_to_count_sql}
+			GROUP BY terms.term_id
+		";
+
+		$results = $wpdb->get_results( $attribute_count_sql ); // phpcs:ignore
+		$results = array_map( 'absint', wp_list_pluck( $results, 'term_count', 'term_count_id' ) );
+
+		$this->set_cache( $transient_key, $results );
+
+		return $results;
+	}
+
+	/**
+	 * Generate calculate query by stock status.
+	 *
+	 * @param string $status status to calculate.
+	 * @param string $product_query_sql product query for current filter state.
+	 * @param array  $stock_status_options available stock status options.
+	 *
+	 * @return false|string
+	 */
+	private function generate_stock_status_count_query( $status, $product_query_sql, $stock_status_options ) {
+		if ( ! in_array( $status, $stock_status_options, true ) ) {
+			return false;
+		}
+		global $wpdb;
+		$status = esc_sql( $status );
+		return "
+			SELECT COUNT( DISTINCT posts.ID ) as status_count
+			FROM {$wpdb->posts} as posts
+			INNER JOIN {$wpdb->postmeta} as postmeta ON posts.ID = postmeta.post_id
+            AND postmeta.meta_key = '_stock_status'
+            AND postmeta.meta_value = '{$status}'
+			WHERE posts.ID IN ( {$product_query_sql} )
+		";
+	}
+
+	/**
+	 * Get the offload filter data.
+	 *
+	 * @param array  $query_vars   The query arguments to calculate the filter data.
+	 * @param string $filter_type The type of filter. Accepts price|stock|rating|attribute.
+	 * @param array  $extra        Some filter types require extra arguments for calculation, like attribute.
+	 */
+	private function pre_get_filter_data( $query_vars, $filter_type, $extra = array() ) {
+		/**
+		 * Allows offloading the filter data.
+		 *
+		 * @hook woocommerce_pre_filter_data
+		 *
+		 * @since 8.7.0
+		 *
+		 * @param array $results      The results for current query.
+		 * @param array $query_vars   The query arguments to calculate the filter data.
+		 * @param string $filter_type The type of filter. Accepts price|stock|rating|attribute.
+		 * @param array $extra        Some filter types require extra arguments for calculation, like attribute.
+		 */
+		return apply_filters( 'woocommerce_pre_filter_data', array(), $query_vars, $filter_type, $extra );
```

Comment:
IMO, ensuring that any data manipulated or provided through hooks is properly sanitized and validated here is an important step, especially considering it involves the input of any user using the filter: WDYT?

Reply:
I agree. Looks like we need a schema validation to ensure the data structure. I think we should do it in a follow-up PR, as this PR is already quite big. What do you think?

Reply:
Sounds good: should we create an issue for this so we don't risk losing track of this change?

Reply:
Here it is: https://github.com/woocommerce/woocommerce/issues/45100